go/kbfs/simplefs/archive.go

```diff
@@ -0,0 +1,876 @@
+	fs := os.DirFS(workspaceDir)
+
+	zipFile, err := os.Create(jobDesc.ZipFilePath)
+	if err != nil {
+		return fmt.Errorf("os.Create(%s) error: %v", jobDesc.ZipFilePath, err)
+	}
+	defer zipFile.Close()
+
+	zipWriter := zipWriterWrapper{zip.NewWriter(zipFile)}
+	defer zipWriter.Close()
+
+	err = zipWriter.AddFS(fs)
+	if err != nil {
+		return fmt.Errorf("zipWriter.AddFS error: %v", jobDesc.ZipFilePath, err)
+	}
+
+	return nil
+}
+
+func (m *archiveManager) zippingWorker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-m.zippingWorkerSignal:
+		}
+
+		jobID, jobCtx, ok := m.startWorkerTask(ctx,
+			keybase1.SimpleFSArchiveJobPhase_Copied,
+			keybase1.SimpleFSArchiveJobPhase_Zipping)
+
+		if !ok {
+			continue
+		}
+		// We got a task. Put another token into the signal channel so we
+		// check again on the next iteration.
+		m.signal(m.zippingWorkerSignal)
+
+		_, _ = jobID, jobCtx
+
+		m.simpleFS.log.CDebugf(ctx, "zipping: %s", jobID)
+
+		err := m.doZipping(jobCtx, jobID)
+		switch err {
+		case nil:
+			m.simpleFS.log.CDebugf(jobCtx, "zipping done on job %s", jobID)
+			m.changeJobPhase(jobCtx, jobID, keybase1.SimpleFSArchiveJobPhase_Done)
+		default:
+			m.simpleFS.log.CErrorf(jobCtx, "zipping error on job %s: %v", jobID, err)
+			m.setJobError(ctx, jobID, err)
+		}
+
+		m.fluseStateFile(ctx)
+	}
+}
+
+func (m *archiveManager) resetInterruptedPhaseLocked(ctx context.Context, jobID string) (changed bool) {
+	switch m.state.Jobs[jobID].Phase {
+	case keybase1.SimpleFSArchiveJobPhase_Indexing:
+		m.simpleFS.log.CDebugf(ctx, "resetting %s phase from %s to %s", jobID,
+			keybase1.SimpleFSArchiveJobPhase_Indexing,
+			keybase1.SimpleFSArchiveJobPhase_Queued)
+		m.changeJobPhaseLocked(ctx, jobID,
+			keybase1.SimpleFSArchiveJobPhase_Queued)
+		return true
+	case keybase1.SimpleFSArchiveJobPhase_Copying:
+		m.simpleFS.log.CDebugf(ctx, "resetting %s phase from %s to %s", jobID,
+			keybase1.SimpleFSArchiveJobPhase_Copying,
+			keybase1.SimpleFSArchiveJobPhase_Indexed)
+		m.changeJobPhaseLocked(ctx, jobID,
+			keybase1.SimpleFSArchiveJobPhase_Indexed)
+		return true
+	case keybase1.SimpleFSArchiveJobPhase_Zipping:
+		m.simpleFS.log.CDebugf(ctx, "resetting %s phase from %s to %s", jobID,
+			keybase1.SimpleFSArchiveJobPhase_Zipping,
+			keybase1.SimpleFSArchiveJobPhase_Copied)
+		m.changeJobPhaseLocked(ctx, jobID,
+			keybase1.SimpleFSArchiveJobPhase_Copied)
+		return true
+	default:
+		m.simpleFS.log.CDebugf(ctx, "not resetting %s phase from %s", jobID,
+			m.state.Jobs[jobID].Phase)
+		return false
+	}
+}
+
+func (m *archiveManager) errorRetryWorker(ctx context.Context) {
+	ticker := time.NewTicker(time.Second * 5)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+
+		func() {
+			m.mu.Lock()
+			defer m.mu.Unlock()
+			jobIDs := make([]string, len(m.state.Jobs))
+			for jobID := range m.state.Jobs {
+				jobIDs = append(jobIDs, jobID)
+			}
+		loopJobIDs:
```

Comment:
I kinda like just putting a label in whenever I need `break` or `continue` in case I add a nested loop somewhere in the future and forget to check for these